[DataGrid] Add `onChange` callback to edit components (#4621)
mui/mui-x: docs/data/data-grid/editing/editing.md

{{"demo": "ConditionalValidationGrid.js", "disableAd": true, "bg": "inline", "defaultCodeOpen": false}}

### Linked fields

The options available for one field may depend on the value of another field.
For instance, if the `singleSelect` column is used, you can provide a function to `valueOptions` returning the relevant options for the value selected in another field, as exemplified below.

```tsx
const columns: GridColDef[] = [
  {
    field: 'account',
    type: 'singleSelect',
    valueOptions: ({ row }) => {

>       return row.type === 'Income' // Gets the value of the "type" field

Reviewer: The row being undefined for the filter, maybe put an example that is compatible with an usage in the filters (to return the two arrays concatenated if row is undefined for instance)

Same on the full example at the end of the section, it crashes when filtering.

____

Side note: 

TS does not complain because in `GridColumns`, by default `R = any` and `any | undefined` is equal to `any`

If I change `R` default value to `{ [key: string]: any }` it does complain and you would have spotted the bug.
I think I put `any` because otherwise you can have type conflicts between typed and untyped `R`, I will try to spend time on it someday.

Author: Updated. For v6 we can change the default type of `R`.

Reviewer: > I think I put any because otherwise you can have type conflicts between typed and untyped R, I will try to spend time on it someday.

Not sure, I did not do it because it was a mess
Maybe there is a way to do it properly but I did not find it

        ? ['Sales', 'Investments', 'Ads']
        : ['Taxes', 'Payroll', 'Utilities'];
    },
  },
];
```

The code above is already enough to display different options in the **Account** column based on the value selected in the **Type** column.

> There is one task left which is to reset the account once the type is changed.

Reviewer: ```suggestion
The only task left is to reset the account once the type is changed.
```

This is needed because, once the type is changed, the previously selected account will not exist in the options.
To solve that, you can create a custom edit component, reusing the built-in one, and pass a function to the `onChange` prop.
This function should call `apiRef.current.setEditCellValue` to reset the value of the other field.

```tsx
const CustomTypeEditComponent = (props: GridEditSingleSelectCellProps) => {
  const apiRef = useGridApiContext();

  const handleChange = async () => {

>     await apiRef.current.setEditCellValue({

Reviewer: I don't know how to improve this
But when reading it for the 1st time, I thought this `onChange` callback was switching the component to a control mode, meaning that users would also have to manually update the value of the current column.

I guess if I understood wrong, other will as well.

Author: Renaming it to `onValueChange` would be better?

Reviewer: Maybe yes
@cherniavskii what's your opinion on this ?

Reviewer 2: @flaviendelangle hmm, I didn't think about controlled prop when reading this, but I get what you mean.
I think `onValueChange` should be better then.

Author: Renamed to `onValueChange`.

      id: props.id,
      field: 'account',
      value: '',
    });
  };

  return <GridEditSingleSelectCell onChange={handleChange} {...props} />;
};
```

The demo below combines the steps showed above.
You can experiment it by changing the value of any cell in the **Type** column.
The **Account** column is automatically updated with the correct options.

{{"demo": "LinkedFieldsRowEditing.js", "disableAd": true, "bg": "inline", "defaultCodeOpen": false}}

> ⚠ The call to `apiRef.current.setEditCellValue` returns a promise that must be awaited.

> > With the `singleSelect` column type, not awaiting for it will cause the other column to be rendered with a `value` that is not in the options.

Reviewer: Not awaiting it before doing what ?
Maybe a small `diff` code could help
